functions/src/download-readings.ts

```diff
@@ -0,0 +1,173 @@
+import * as functions from 'firebase-functions';
+import * as fs from 'fs';
+import * as path from 'path';
+import * as os from 'os';
+import * as admin from 'firebase-admin';
+import SensorReading from './sensor-reading';
+import {firestore} from './admin';
+
+/**
+ * Uploads a file with name filename and specified data to the default storage
+ * bucket of the Firebase project.
+ * @param filename - name of file to be uploaded to Firebase bucket
+ * @param data - data to write to file
+ */
+function uploadFileToFirebaseBucket(filename: string, data: string) {
+  const tempLocalFile = path.join(os.tmpdir(), filename);
+  return new Promise<void>((resolve, reject) => {
+    // Write data into the temp file
+    fs.writeFile(tempLocalFile, data, error => {
+      if (error) {
+        reject(error);
+        return;
+      }
+
+      // Upload file into current Firebase project default bucket
+      admin
+        .storage()
+        .bucket()
+        .upload(tempLocalFile)
+        .then(() => resolve())
+        .catch(error => reject(error));
+    });
+  });
+}
+
+/**
+ * @param message - PubSub message that includes JSON with start and end fields
+ *
+ * @remarks
+ * The start and end values represent the non-inclusive interval of time to fetch
+ * readings for. The start and end values should be numbers that are the number
+ * of milliseconds since EPOCH.
+ *
+ * @example
+ * ```
+ * // Gets readings for the month of December (in PST)
+ * generateReadingsCsv(
+ *   {data: Buffer.from('{"start": 1606809599000, "end": 1609488000000}')}
+ * )
+ * ```
+ */
+async function generateReadingsCsv(
+  message?: functions.pubsub.Message
+): Promise<void> {
+  let startMilliseconds: number | undefined = undefined;
+  let endMilliseconds: number | undefined = undefined;
+
+  // Get the `start` and `end` attributes of the PubSub message JSON body.
+  if (message) {
+    try {
+      startMilliseconds = message.json.start;
+      endMilliseconds = message.json.end;
+    } catch (error) {
+      // No error, use default values
+    }
+  }
+
+  let startDate: Date | undefined = undefined;
+  let endDate: Date | undefined = undefined;
+
+  // Validate start and end dates, or set defaults
+  if (
+    startMilliseconds &&
+    endMilliseconds &&
+    startMilliseconds < endMilliseconds
+  ) {
+    startDate = new Date(startMilliseconds);
+    endDate = new Date(endMilliseconds);
+  } else {
+    // Default to the most recent month
+    startDate = new Date();
+    endDate = new Date(startDate);
+    endDate.setMonth(startDate.getMonth() - 1); // eslint-disable-line no-magic-numbers
+  }
+
+  // Initialize csv with headers
+  const headings = SensorReading.getCsvHeader();
+
+  const sensorList = (await firestore.collection('/sensors').get()).docs;
+  const readingsArrays = new Array<Array<string>>(sensorList.length);
```

Comment:
I can't remember, have we been using this notation or string[][] notation?

Reply:
We have used both throughout. Do you have an opinion about which to use?

Reply:
Not really, I just wasn't sure if we had previously only been using one, we would want consitency but we are already inconsistent so 🤷‍♂️